// common/responseTool.js
var responseTool = {
    tool: require('./tool'),
    jsdom: require('../public_nodejs/jsdom').jsdom,
    fs: require('fs'),
    /*返回服务端异常*/
    sendServerException: function (res, mess) {
        res.send(500, mess || '服务端错误请查看后台日志');
    },
    /*返回拒绝访问*/
    sendDecline: function (res, mess) {
        res.send(405, mess || '服务端拒绝访问');
    },
    /*返回数据更新、上传、下载成功*/
    sendSuccess: function (res, param) {
        param = param || {};
        res.set('Content-Type', 'text/json;charset=utf-8');
        res.send(param);
    },
    /*重定向*/
    sendRedirect: function (res, url) {
        res.send(302, url);
    },
    /*登录请求的响应
     */
    loginResponse: function (req, res, result) {
        var err = result.err;
        if (err) {
            console.error('登录err：' + (err.stack || JSON.stringify(err)));
            return this.renderLogin(res, 1);
        }
        this.tool.storeUserInfo(req, res, result.puser);
        this.redirectMain(res);
    },
    /*渲染主页面*/
    renderMain: function (res, userInfo, ticket) {
        var mainPath = './views/';
        var mainHtml = 'layout.html';
        var mainHtmlPath = mainPath + mainHtml;
        var newMainHtmlName = '_layout.html';
        var newMainHtmlPath = mainPath + newMainHtmlName;
        var _this = this;
        this.jsdom.env({
            file: mainHtmlPath,
            done: function (err, window) {
                var document = window.document;
                var newHeadElement = document.createElement('script');
                newHeadElement.setAttribute('type', 'text/javascript');
                newHeadElement.innerHTML = 'var ' + pconst.pticket + '="' + ticket + '";';
                var insertBeforeNode = document.getElementsByTagName('title')[0];
                document.head.insertBefore(newHeadElement, insertBeforeNode);

                var bodyHtml = document.documentElement.outerHTML.replace(/(&lt;)/g, '<').replace(/(&gt;)/g, '>');
                _this.fs.writeFile(newMainHtmlPath, bodyHtml, function (writeErr) {
                    if (writeErr) {
                        console.log('写入layout页面时出错：' + writeErr.stack || (JSON.stringify(writeErr)));
                        return _this.sendServerException(res);
                    }
                    res.render(newMainHtmlName, { host: commonLibUrl, user: userInfo });
                });
            }
        });
    },
    /*渲染登录页面*/
    renderLogin: function (res, err) {
        res.render('login', { host: commonLibUrl, err: err || '' });
    },
    /*跳转到首页*/
    redirectMain: function (res) {
        res.redirect('/');
    },
    /*无权限访问*/
    renderNoPower: function (res, isLogin) {
        if (isLogin === true) return this.renderLogin(res);
        this.sendDecline(res, '无权限访问');
    }
};

module.exports = responseTool;
